### PR TITLE
Сlstm refactoring

### PR DESCRIPTION
**Pull request by valer1435**
Some changes that can increase efficiency of clstm algorithm. Also bug with composer's inability to use model during composing was solved.

But using such heavy model in composing still questionable due to low performance compared with it's weight (maybe it can be useful on some types of data and with small horizon but I can't say exactly). Maybe we can refactor model structure, advices are welcome.

### CODE REVIEW

File: examples/simple/time_series_forecasting/clstm.py [hunk 1]
```diff
@@ -41,60 +45,66 @@ def get_ts_data_long(n_steps=80, forecast_length=5):
...
+        'num_epochs': 50}).to_pipeline()
+
+    #pipeline = PipelineBuilder().add_node("lagged", params={'window_size': window_size}).add_node("ridge").to_pipeline()
+    # pipeline_tuner = TunerBuilder(train_data.task) \
+    #     .with_tuner(PipelineTuner) \
+    #     .with_metric(RegressionMetricsEnum.RMSE) \
+    #     .with_cv_folds(2) \
+    #     .with_validation_blocks(1) \
+    #     .with_iterations(2).build(train_data)
+    # pipeline_tuner.tune(pipeline)
```
Comment (nicl-nno) [inline examples/simple/time_series_forecasting/clstm.py:69]: Почисти сразу закомментированный код.

**Comment by valer1435** [comment]
Из основных изменений - убрал lagged преобразование из модели и заменил последовательное предсказание интервальным

**Comment by aPovidlo** [comment]
А `'learning_rate': 0.0004,` при `'num_epochs': 50` не слишком ли мал?

**Review by staeros — APPROVED** [review]

File: examples/simple/time_series_forecasting/clstm.py (hunk 1 above)
Comment (valer1435) [reply examples/simple/time_series_forecasting/clstm.py]: [Убрал](https://github.com/aimclub/FEDOT/pull/1029/commits/351c92b935080224d2e4a1acfc0c27b3923ea45e#diff-98ab2e857d2a8be66f8beb3439f8604a919d6128b9122f2804ca1528ece87954R65-R66)

File: examples/simple/time_series_forecasting/api_forecasting.py [hunk 2]
```diff
@@ -20,10 +19,9 @@
     'stackoverflow': f'{fedot_project_root()}/examples/data/ts/stackoverflow.csv'}
 
 
-def run_ts_forecasting_example(dataset='australia', horizon: int = 30, validation_blocks=2, timeout: float = None,
-                               visualization=False):
+def get_ts_data(dataset: str = 'beer', horizon: int = 10, validation_blocks: int = 1):
     time_series = pd.read_csv(datasets[dataset])
-
+    #time_series['value'] = (time_series['value'] - np.mean(time_series['value'])) / np.std(time_series['value'])
```
Comment (nicl-nno) [inline examples/simple/time_series_forecasting/api_forecasting.py:24]: Закомментированный код всё ещё остался.

File: fedot/core/pipelines/verification_rules.py [hunk 3]
```diff
@@ -105,7 +105,8 @@ def is_pipeline_contains_ts_operations(pipeline: Pipeline):
 
 def has_no_data_flow_conflicts_in_ts_pipeline(pipeline: Pipeline):
     """ Function checks the correctness of connection between nodes """
-
+    if 'cgru' in pipeline.descriptive_id:
+        print('cgru')
```
Comment (nicl-nno) [inline fedot/core/pipelines/verification_rules.py:109]: Не нужно?

File: examples/simple/time_series_forecasting/ts_pipelines.py [hunk 4]
```diff
@@ -242,24 +242,24 @@ def ts_naive_average_ridge_pipeline():
...
 
-    Where clstm - convolutional long short-term memory model
+    Where cgru - convolutional long short-term memory model
     """
-    clstm_params = {'window_size': 29, 'hidden_size': 50, 'learning_rate': 0.004,
+    cgru_params = {'window_size': 29, 'hidden_size': 50, 'learning_rate': 0.004,
                     'cnn1_kernel_size': 5, 'cnn1_output_size': 32,
                     'cnn2_kernel_size': 4, 'cnn2_output_size': 32,
                     'batch_size': 64, 'num_epochs': 3, 'teacher_forcing': 0.8,
                     'optimizer': 'adam', 'loss': 'mse'}
```
Comment (nicl-nno) [inline examples/simple/time_series_forecasting/ts_pipelines.py:258]: А всё это обязательно указывать? Если нет, то хорошо бы в одном из примеров продемонстировать вариант без явно заданных гиперпараметров.

File: test/unit/tasks/test_forecasting.py [hunk 5]
```diff
@@ -312,21 +312,19 @@ def test_clstm_forecasting():
...
-def test_clstm_in_pipeline():
+def test_cgru_in_pipeline():
     horizon = 5
     n_steps = 100
     train_data, test_data = get_ts_data(n_steps=n_steps + horizon, forecast_length=horizon)
 
-    pipeline = clstm_pipeline()
+    pipeline = cgru_pipeline()
     pipeline.fit(train_data)
     predicted = pipeline.predict(test_data).predict[0]
```
Comment (nicl-nno) [inline test/unit/tasks/test_forecasting.py:329]: А тут время выполнения теста норм, не слишком долго обучается cgru?

**Comment by nicl-nno** [comment]
За рамками PR - надо будет проработать схему реализации таких тяжелых моделей на базе F.Ind. 
Потому что тут мы все равно в стандарные пресеты lstm-ы не включаем и о их существовании мало кто знает.

File: docs/source/basics/ts_forecasting.rst [hunk 6]
```diff
@@ -629,7 +629,7 @@ Examples
 **Simple**
 
 * `Forecasting with FEDOT API <https://github.com/nccr-itmo/FEDOT/blob/master/examples/simple/time_series_forecasting/api_forecasting.py>`_
-* `Forecasting using clstm <https://github.com/nccr-itmo/FEDOT/blob/master/examples/simple/time_series_forecasting/clstm.py>`_
+* `Forecasting using cgru <https://github.com/nccr-itmo/FEDOT/blob/master/examples/simple/time_series_forecasting/cgru.py>`_
```
Comment (aPovidlo) [inline docs/source/basics/ts_forecasting.rst:632]: Доки может тоже изменить и прочие картинки, т.к. ты же все-равно оставил LSTM

File: fedot/core/operations/evaluation/operation_implementations/models/ts_implementations/cgru.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+        self.model = CGRUNetwork(
+            hidden_size=int(params.get("hidden_size")),
+            cnn1_kernel_size=int(params.get("cnn1_kernel_size")),
+            cnn1_output_size=int(params.get("cnn1_output_size")),
+            cnn2_kernel_size=int(params.get("cnn2_kernel_size")),
+            cnn2_output_size=int(params.get("cnn2_output_size"))
+        )
+
+        self.optim_dict = {
+            'adam': torch.optim.AdamW(self.model.parameters(), lr=self.learning_rate),
```
Comment (aPovidlo) [inline fedot/core/operations/evaluation/operation_implementations/models/ts_implementations/cgru.py:39]: `Adam` и `AdamW` это разные оптимизаторы. Чтобы не было путаницы у пользователей, лучше добавить в словарь отдельный ключ adamw.

File: fedot/core/operations/evaluation/operation_implementations/models/ts_implementations/cgru.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+        Implementation uses the idea of teacher forcing. That means model learns
+        to predict data when horizon != 1. It uses real values or previous model output
+        to predict next value. self.teacher_forcing param is used to control probability
+        of using real y values.
+
+        :param train_data: data with features, target and ids to process
+        """
+        self.model.init_linear(train_data.task.task_params.forecast_length)
+        self.model = self.model.to(self.device)
+        data_loader = self._create_dataloader(train_data)
```
Comment (aPovidlo) [inline fedot/core/operations/evaluation/operation_implementations/models/ts_implementations/cgru.py:68]: Еще можно было бы добавить для улучшения обучения динамическое изменение learning_rate от валидационных данных [lr_scheduler](https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate)

File: fedot/core/operations/evaluation/operation_implementations/models/ts_implementations/cgru.py [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+        )
+        self.gru = nn.GRU(cnn2_output_size, self.hidden_size, dropout=0.1)
+        self.hidden_cell = None
+        self.linear = None
+
+    def init_linear(self, forecast_length):
+        self.linear = nn.Linear(self.hidden_size, forecast_length)
+
+    def init_hidden(self, batch_size, device):
+        self.hidden_cell = torch.zeros(1, batch_size, self.hidden_size).to(device)
```
Comment (aPovidlo) [inline fedot/core/operations/evaluation/operation_implementations/models/ts_implementations/cgru.py:182]: Вроде если какие-то случайные веса генерировать, то это может быть более профитнее, чем просто от нулей отталкиваться.

File: examples/simple/time_series_forecasting/cgru.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+                     features=time_series,
+                     target=time_series,
+                     task=task,
+                     data_type=DataTypesEnum.ts)
+    return train_test_data_setup(data), task
+
+
+def cgru_forecasting():
+    """ Example of cgru pipeline serialization """
+    horizon = 2
```
Comment (aPovidlo) [inline examples/simple/time_series_forecasting/cgru.py:48]: А не слишком ли малый горизонт?

**Review by aPovidlo — CHANGES_REQUESTED** [review]
Не совсем понял, `clstm` остается в коде, но при это во многих местах о нем информация удалилась или была заменена на `cgru`. Может быть стоит восстановить везде про `lstm`?
Потенциально, в пайплайне их можно же обе использовать.

File: examples/simple/time_series_forecasting/cgru.py (hunk 10 above)
Comment (valer1435) [reply examples/simple/time_series_forecasting/cgru.py:48]: На маленьком горизонте он побеждает ридж

File: examples/simple/time_series_forecasting/api_forecasting.py (hunk 2 above)
Comment (valer1435) [reply examples/simple/time_series_forecasting/api_forecasting.py]: [Убрал](https://github.com/aimclub/FEDOT/pull/1029/commits/9d1430cf2a504fbb4c0073c722892ceba4dde8dd#diff-61d9c62b97ba3e153b9fb93d8340113be9aeebf42be709a8b91dc56fa2c04ab9R22-R23)

File: fedot/core/pipelines/verification_rules.py (hunk 3 above)
Comment (valer1435) [reply fedot/core/pipelines/verification_rules.py]: [Нет](https://github.com/aimclub/FEDOT/pull/1029/commits/9d1430cf2a504fbb4c0073c722892ceba4dde8dd#diff-626f6b72c51c57024482629cd3ec367c301efd52b2e8883b75c2c6ce394c74eaR107-R108)

File: examples/simple/time_series_forecasting/ts_pipelines.py (hunk 4 above)
Comment (valer1435) [reply examples/simple/time_series_forecasting/ts_pipelines.py]: [Убрал](https://github.com/aimclub/FEDOT/pull/1029/commits/9d1430cf2a504fbb4c0073c722892ceba4dde8dd#diff-98ab2e857d2a8be66f8beb3439f8604a919d6128b9122f2804ca1528ece87954R13-R16)

File: test/unit/tasks/test_forecasting.py (hunk 5 above)
Comment (valer1435) [reply test/unit/tasks/test_forecasting.py]: Действительно, долго. 6-7 секунд на 2 теста. Раньше посчитали как-то не критичным. Убрал в [интеграционные](https://github.com/aimclub/FEDOT/pull/1029/commits/9d1430cf2a504fbb4c0073c722892ceba4dde8dd#diff-341077bd875805ca4266bfae2598d75c35ee5f0260ba3e473aef318e29818b16R7)

File: fedot/core/operations/evaluation/operation_implementations/models/ts_implementations/cgru.py (hunk 7 above)
Comment (valer1435) [reply fedot/core/operations/evaluation/operation_implementations/models/ts_implementations/cgru.py]: [Поменял](https://github.com/aimclub/FEDOT/pull/1029/commits/9d1430cf2a504fbb4c0073c722892ceba4dde8dd#diff-2e757189efed106001507ee9fb93df8a34d81ecec4728aaddf402c28b5c44ea7R166)

File: fedot/core/operations/evaluation/operation_implementations/models/ts_implementations/cgru.py (hunk 8 above)
Comment (valer1435) [reply fedot/core/operations/evaluation/operation_implementations/models/ts_implementations/cgru.py:69]: Не помогло, даже хуже стало. А подбирать параметры еще и scheduler'a в автомле будет уже слишком.

File: fedot/core/operations/evaluation/operation_implementations/models/ts_implementations/cgru.py (hunk 9 above)
Comment (valer1435) [reply fedot/core/operations/evaluation/operation_implementations/models/ts_implementations/cgru.py]: Насчет профита не знаю, вроде и нулями норм заполнять. Но в доках торча, в примере они заполняют случайно, да.
[Обновил](https://github.com/aimclub/FEDOT/pull/1029/commits/9d1430cf2a504fbb4c0073c722892ceba4dde8dd#diff-4d20c87c08c249746ab5e12a4b89e271258cc14a133e37fe2125f623e55833a9R183)

File: examples/simple/time_series_forecasting/cgru.py (hunk 10 above)
Comment (valer1435) [reply examples/simple/time_series_forecasting/cgru.py]: [Теперь 12](https://github.com/aimclub/FEDOT/pull/1029/commits/9d1430cf2a504fbb4c0073c722892ceba4dde8dd#diff-98ab2e857d2a8be66f8beb3439f8604a919d6128b9122f2804ca1528ece87954R11). На самом деле не принципиально

**Review by aPovidlo — APPROVED** [review]

**Review by nicl-nno — APPROVED** [review]

**Review by staeros — APPROVED** [review]
